### DISCUSSION TITLE
Upgrade from cppunit to catch2

### DISCUSSION THREAD

**Discussion by simsong** [sleuthkit/sleuthkit#2950] [General]
We had really good luck on the bulk_extractor project using catch.
https://github.com/catchorg/Catch2/blob/master/docs/tutorial.md

My intent is to create a similar test framework on TSK. I'm looking for comments and suggestions.

**Comment by uckelman-sf** [comment]
We've been using Catch2 for years on various projects, including lightgrep. I'd happily set it up for the tests in #1057 after that PR is merged.

**Comment by simsong** [comment]
How would you set it up?

**Reply by uckelman-sf** [reply]
The same as in lightgrep:

* https://github.com/strozfriedberg/lightgrep/blob/main/configure.ac#L316
* https://github.com/strozfriedberg/lightgrep/blob/main/Makefile.am#L342

Tests are in `test/.

**Comment by simsong** [comment]
Apparently it is set up to use cppunit. However, it wasn't installed on my system. How does this system declare its requirements and prepare VMs? Should we create scripts to do that?

**Reply by uckelman-sf** [reply]
Hmm. It looks like nothing is ensuring cppunit is there in CI. Given that we aim to stop using it immanently, it's a waste of time to get cppunit working.

I'll switch #1057 over to Catch2 next.

**Comment by simsong** [comment]
Why do we want to stop using it immediately? Most of the dev crew at SKL is using Java, so I figured that they picked cppunit back in 2010 to make everybody feel happy. I have it working on my mac right now. I'm trying to make as few changes as possible in the code base at the moment.

**Reply by uckelman-sf** [reply]
If you look in `unit_tests/` in develop, you'll see that there are a total of three tests using cppunit, none of which test anything of substance in Sleuthkit. The conclusion I reached from this is that someone set up cppunit about 15 years ago and then didn't use it.

I.e., we should stop using cppunit before anyone seriously starts using it.

**Comment by simsong** [comment]
True, but I have it working.

**Reply by simsong** [reply]
If there is a significant reason to move to catch, we can do that. But right now I want to concentrate on new stuff. Is there a serous reason to avoid cppunit?

**Reply by uckelman-sf** [reply]
Given that we're starting from close to zero with unit tests, we should chose the best test framework we can. 

I rarely encounter projects using cppunit these days. It's unpleasant to work with. I'm not enthusiastic about writing tests with it and I expect other people who might contribute tests would feel the same way.

Catch2 is pleasant to use. We've both had good experiences with it. I'm volunteering to set it up. Catch2 has support for test order randomization, which helps flush out otherwise hard to find bugs; test sharding, so we can easily parallelize tests; and benchmarking, which could help us make Sleuthkit faster. I've used all these features and would miss them in cppunit.

Writing new tests with cppunit when we aim eventually to convert to Catch2 is wasting all the time we will spend converting those tests. We could write them in the better framework from the start and spend less time doing it.

**Comment by simsong** [comment]
You convinced me. I’ll move over in a few days.

**Comment by uckelman-sf** [comment]
I've begun in PR #2956.

**Comment by simsong** [comment]
@uckelman-sf  - Catch has undergone a significant change between v2 and v3. v2 is a single header to include; v3 now is a full library, with lots of headers. It is on the devel branch. 

https://github.com/catchorg/Catch2/

I'm hesitant to go with catch v3 because it is still devel, but migrating from v2 to v3 is not trivial. What is your recommendation?

**Reply by uckelman-sf** [reply]
All of our internal projects are already using v3. I'd prefer to avoid migrating altogether and start with v3 (which is what I've done already in the PR).

**Reply by simsong** [reply]
Do you prefer v3 to be embedded directly, or to be a submodule, or to be separately installed?

**Reply by uckelman-sf** [reply]
Internally, our overall CI framework builds Catch2. For CI here, I'm not sure yet what will be simplest. I'll look into options for that next. (Not until mid-week, though. I'm away for the next few days.)

**Comment by simsong** [comment]
Homebrew can install catch
https://formulae.brew.sh/formula/catch2

Looks like their are similar packages for Ubuntu

**Reply by uckelman** [reply]
If we can get it as a package, that saves us from needing to build it. That's the best approach. I'm glad it works.

**Reply by simsong** [reply]
It works. I hard-coded the cppunit to make it work on homebrew, but I'm eager to get rid of it. It's all yours now.
Once you have catch deployed (I guess we are going with catch2, since catch3 is under devel still), I'll take a closer look at the codecov mess.

**Comment by simsong** [comment]
Looks like we've done this.
